File: server/src/entities/Category.ts
import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { PrefixRule } from './PrefixRule';
import { Transaction } from './Transaction';

export type CategoryType = 'income' | 'expense' | 'transfer';

@Entity()
export class Category {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({
    unique: true,
  })
  name: string;

  @Column({ nullable: true })
  color: string;

  @OneToMany(() => PrefixRule, (rule) => rule.category, { cascade: true })
  prefixRules: PrefixRule[];

  @OneToMany(() => Transaction, (transaction) => transaction.upload)
  transactions: Transaction[];

  @ManyToOne(() => Category, (category) => category.subcategories, {
    nullable: true,
  })
  parentCategory: Category;

  @Column('text', {
    nullable: false,
    default: 'expense',
  })
  type: CategoryType;

  @Column('int', { nullable: true })
  parentCategoryId: number;

  @OneToMany(() => Category, (category) => category.parentCategory)
  subcategories: Category[];
}

File: go-server/model/category.go
package model

import (
	"gorm.io/gorm"
)

type Category struct {
  gorm.Model
	Name string `gorm:"unique;not null"`
	Color *string
	IconURL *string
	Transactions []Transaction
	Type string `gorm:"default:expense;not null"`
	ParentCategoryID *uint
  SubCategories      []Category `gorm:"foreignkey:ParentCategoryID"`
	PrefixRules []PrefixRule
}